### DISCUSSION TITLE
Continous touch event for InputContainer?

### DISCUSSION THREAD

**Discussion by sethhammett** [koreader/koreader#8763] [Q&A] [answered]
So recently I've been doing some plugin development for KOReader, and one of the things I'm making is a drawing program. The drawing canvas is an InputContainer, but all the events available only seem to fire when you first touch or release the screen. I'm looking for an event that fires every time your finger moves, so that the user can draw continuous lines on the canvas. Is there such an event, and if so, how would I implement it?

Here is the plugin so far (notice how I can only draw discrete points):
![Reader_2022-02-01_161214](https://user-images.githubusercontent.com/83931313/152069247-c57f9566-c435-4df8-a273-3cc31e85b501.png)

**Comment by NiLuJe** [comment]
IIRC, neither the Input nor Gesture modules are made for that, they're fairly specifically designed around UI-effective gestures only ;).

I vaguely recall that someone had made attempts at dealing with pen drawing on the rM, so all hope is not lost ;).

**Reply by sethhammett** [reply]
Ok thanks! I'll keep looking at the KOReader code, I might find some way around that.

**Reply by NiLuJe** [reply]
There may have been mention of a custom inputHandler (e.g., look at frontend/device/input)

**Reply by Frenzie** [reply]
Link dump

https://github.com/koreader/koreader/issues/2633

Possibly see https://github.com/koreader/koreader/issues/2633#issuecomment-749779818

Also:
https://github.com/koreader/koreader/issues/8537

**Comment by sethhammett** [comment, marked as answer]
After learning that InputContainer did not have this feature, I looked at `./frontend/device/input.lua` per NiLuJe's suggestion to try and find a way to get the touch coordinates without fancy events and stuff, and I found these!
```
self:setCurrentMtSlot("x", ev.value)
self:getCurrentMtSlotData("tool")
```
So I extrapolated from there and came up with these:
```
Device.input:getCurrentMtSlotData("x") -- Get touch X coordinate
Device.input:getCurrentMtSlotData("y") -- Get touch Y coordinate
Device.input:getCurrentMtSlotData("id") -- If id is -1 then not touching, otherwise touching
```
I decided to make my plugin run a function (self:doDrag()) and then that function checks if the user is still touching the screen. and if they are it runs itself after 0.05 seconds, thus creating a loop that only runs when the user is touching the screen. I don't know if anyone other than me would actually use these functions but I just thought it was interesting. Anyways here's the code:
```
[...]

function Canvas:init()
    [...]

    self.drag_callback = function () self:doDrag() end

    self.ges_events = {
        TouchCanvas = {
            GestureRange:new{
                ges = "touch",
                range = function () return self.dimen end,
            },
        },
    }

    [...]
end

[...]

function Canvas:onTouchCanvas()
    self._touch_x = nil
    self._touch_y = nil

    self:doDrag()
end

function Canvas:doDrag()
    [...]
    
    self._touch_x = math.floor(Device.input:getCurrentMtSlotData("x") / self._zoom)
    self._touch_y = math.floor(Device.input:getCurrentMtSlotData("y") / self._zoom)

    [...]

    self._image:paintCircle(self._touch_x, self._touch_y, 2, Blitbuffer.COLOR_BLACK)
    self:_update(self._touch_x - 2, self._touch_y - 2, 5, 5)

    if Device.input:getCurrentMtSlotData("id") ~= -1 then
        UIManager:scheduleIn(0.05, self.drag_callback)
    end
end

[...]
```
And here's a smiley face! (notice the smooth lines)
![FileManager_2022-02-01_194906](https://user-images.githubusercontent.com/83931313/152087923-c782fc84-88dc-41da-916a-5ee71590cc88.png)
